### ISSUE TITLE
Move HEIC support to extension

### ISSUE THREAD

**Issue by wjt** [flathub/org.gimp.GIMP#111]
`org.gimp.GIMP` includes HEIF/HEIC support, which links to libx265 and libde265. Some distributors or users may prefer not to have such codecs installed on their system.

Other apps like Totem, Pitivi, Blender, … build a separate `$APP.Codecs` extension, which is automatically installed by default but can be masked by the end-user or distributor to avoid it being installed. They define the extension in their manifest like this:

```json
    "add-extensions": {
        "org.pitivi.Pitivi.Codecs": {
            "directory": "lib/codecs",
            "add-ld-path": "lib",
            "bundle": true,
            "autodelete": true
        }
    }
```

and arrange for (sub)directories of `/app/lib/codecs` to be added to their library and plugin paths, e.g.:

```json
    "finish-args": [
        ...,
        "--env=GST_PLUGIN_PATH=/app/lib/codecs/lib/gstreamer-1.0",
        ...
    ],
    ...
    "build-options": {
        "env": {
            "GST_PLUGIN_SYSTEM_PATH": "/app/lib/gstreamer-1.0/:/app/lib/codecs/lib/gstreamer-1.0/",
            ...
        },
        "append-ld-library-path": "/app/lib/codecs/lib",
        "ldflags": "-L/app/lib/codecs/lib"
    },
```

Then add `post-install` commands to the relevant modules in the Flatpak to move them to subdirectories of `/app/lib/codecs`.

Might it be possible to give GIMP the same treatment? From a brief investigation it seems that `/app/lib/gimp/2.0/plug-ins/file-heif/file-heif` is already a stand-alone module and that other modules do not link to x265/de265 so it may be not too difficult to move it, and the libraries it depends on, to an extension. 

The one downside of this approach is that it is not possible to update the `MimeType` list in the `.desktop` file based on whether the extension is installed. These apps choose to always list the relevant formats. I think this is acceptable – by not installing the extension the user/distributor is deliberately choosing a less-functional app.

**Comment by Jehan** [comment]
> a separate $APP.Codecs extension, which is automatically installed by default but can be masked by the end-user or distributor to avoid it being installed.

Sure with such conditions, it looks fair.

> From a brief investigation it seems that /app/lib/gimp/2.0/plug-ins/file-heif/file-heif is already a stand-alone module and that other modules do not link to x265/de265 so it may be not too difficult to move it, and the libraries it depends on, to an extension.

Indeed.

So there is still a detail, which is that libheif provides both AVIF and HEIC support. AVIF is usually not a problem at all (because the problem with HEIC is usually a patent problem and AVIF has been created especially not to have such a patent). So now we have a runtime check of the available decoders and encoders available.

First this means that the same `plug-ins/file-heif/file-heif` code will work with or without HEIC support. Second it means that we must always include this plug-in, but simply not include the encoder/decoder part in the main flatpak. The extension must only add these to be foundable by the plug-in at runtime. This way, even when masking this extension, people won't lose AVIF support.

In any case, sure this seems a completely fair proposition. I would definitely accept a patch/merge request proposing such change. You seem to know how to do it so be my guest! 🤗

**Comment by wjt** [comment]
Ah, I had missed that detail. It makes it a bit more tricky compared to just moving some files into an extension. file-heif links to libheif which links to x265/de265, so if the latter are missing the plugin will presumably fail to load. So I guess one would either have to make libheif dlopen those libraries and cope with their absence; or libheif could be compiled twice, once with and once without 265 support, and put the former copy in a `.Codecs` extension and the latter in the main Flatpak.

I won't have the cycles to work on this at present I'm afraid. Perhaps some day!

**Comment by wjt** [comment]
I discovered that libheif supports building codecs as dynamically-loaded plugins, which is just what we need here. In short, this becomes a matter of building it with `-DWITH_LIBDE265_PLUGIN=ON -DWITH_X265_PLUGIN=ON` and then rearranging files and library paths to split off the offending pieces. I am several hours into a test build of GIMP but the intermediate artefacts look good: libheif.so no longer links directly to either x265/de265.

**Comment by Jehan** [comment]
> I discovered that libheif supports building codecs as dynamically-loaded plugins, which is just what we need here. 

That was indeed what I meant in my previous comment:

> Second it means that we must always include this plug-in, but simply not include the encoder/decoder part in the main flatpak. The extension must only add these to be foundable by the plug-in at runtime. This way, even when masking this extension, people won't lose AVIF support.

Sorry if that was not clear!

**Comment by Jehan** [comment]
This got implemented as #262. Closing! :-)

**Comment by brunvonlope** [comment]
Ported to upstream: https://gitlab.gnome.org/GNOME/gimp/-/commit/ebeef5d7ef94273005436efccdb742a64b617d3d
